### ISSUE TITLE
Suggestion :: Design Choice :: Central Server Infastructure Elimination

### ISSUE THREAD

**Issue by FrantzelasG** [warner/magic-wormhole#254]
With 1 little extra step, we can eliminate the need for a central server. We can easily send our public ip address using email,sms,phone,etc using hexadecimal digits, like this: 192.168.1.1 -> C0A80101. It's well within humans ability to remember 8 chars, if for example you own a static IP, you can memorize your IP, like you do with your passwords. Using base-62(26x2+10) encoding we can go down to 6 chars transmitted. The library should be able to identify computer's public ip, encode it properly and echo it to user. For example user executes "worhole send thisfile*" , app prints "AzS9sF".

**Comment by warner** [comment]
Interesting idea.. thanks!

You'll still need the secret words though (to protect against network adversaries, or the server being compromised), so the code would be at least `wormhole receive AzS9sF-purple-sausages`. And we'd have to either require the server to run on a well-known port, or include the port number too (`wormhole receive AzS9Sf:4000-purple-sausages`). And I'd like this protocol to survive the transition to IPv6 (`wormhole receive 2600:3c01::e8f3:9145:fed4:59ed:4000-purple-sausages`), which makes the server addresses much to long to feasibly transcribe.

Also, base62 is maximal for things that need to be typed, but since the codes need to be verbally transcribable too, I don't think we can afford to have upper-vs-lowercase. So base32 or base36 is probably the densest we can do, which makes the encoded IP addresses slightly larger. A big part of the value of magic-wormhole is how easy the codes are to speak and to write, which constrains what we can do with the server address.

**Comment by FrantzelasG** [comment]
0) Protocols use std ports since ever. Unless security is a concern user should be permitted to either use a standard port and reduce typing load or use a random port and transmit it. SSH faced similar problems, when in fact stakes where much higher, a rooted machine vs 1 file.
1) Base36 is indeed easier to transmit verbally. Again user could choose what to use. At library side, library could calculated both. At receive side typing either, can be easily detected. 
2) Maybe sender/receiver could opt for 1-time passwords, or a pre-shared key/secret knock
3) ipv6 is a tough nut. 128 bits ~ 25 chars/base36. Static Ip & caching could help.  Maybe static-ip are issued by default in a ipv6 world. Cheers!

**Comment by bj0** [comment]
couldn't you use a DHT to find each other?

**Comment by warner** [comment]
@bj0 yeah, probably, but we need to allocate unique channel IDs, which is a lot easier with a centralized service.. it can probably be done reliably-enough in a DHT, but would take more work and more time, increasing latency. Let's use #72 to brainstorm about DHT options.

@FrantzelasG I forgot to mention NAT.. most of the time, when I'm sending a file to someone else, neither of us has a public IP address, so we need *something* that we can both connect to, just to bootstrap the hole-punching process. And I've found that most of my machines that *do* have a publically-reachable IP address aren't really personal machines, they're servers, and they've got stable public hostnames too, and I just remember the hostname.

I agree that using standard ports is fine, so we needn't worry about fitting those extra characters into the string.

Using base36 (or 62 or whatever) loses a lot of the error-checking options for the code: any typo will prevent the connection from being made. (In the current `4-purple-sausages` scheme, the "4" is critical and has no error-checking, but the "purple" and "sausages" come from a short wordlist and so the receiving client could use the normal spell-checking techniques to find the closest word and tolerate typos). So I'd like to make the non-error-checked portion as short as possible.

The `purple-sausages` part of the existing scheme is basically a one-time password (although the PAKE crypto we use makes it safe even if someone guesses the code afterwards). Using a pre-shared key or some out-of-band thing like port-knocking means you have to deliver that secret along with the rest of the code, making it bigger and more error-prone (e.g. "send me a file, by typing in IP address X.Y.Z.Q, and also make a TCP connection to port 1948 just before you run the program, and also pre-encrypt the file with key abc..256..bits..later..xyz", vs "4-purple-sausages"). And, if it's optional, then people will prefer to not use it, and then this isn't a secure file-transfer tool anymore.

My other concern with transcribing an IP address is that the direction of the file transfer might not align with the direction of the information transfer. If I'm sending a file to you, and you've got a public IP address (but I don't), then first you have to tell me your IP address, and then I send you the file. That's backwards: if I'm sending the file to you, then I want to be telling you some string, not the other way around. They might line up if I happen to have a public IP (then I tell you "please contact this IP and ask for a file") but there's no guarantee that it'll work out that way.

And yeah, v6 makes all manual-adress-handling things harder. I'm glad there are folks doing things like automatic-local-service-discovery (MDNS-type stuff), that's going to make things a lot better.

thanks!

**Comment by drequivalent** [comment]
We don't need to transfer the IP address of the user.
Instead, we can have a network of rendez-vous servers, that know about each other (hold a list of other servers, join new servers to this list).
A client bootstraps into this network and downloads a list of servers. Then, it picks one at random, performs the usual rendez-vous stuff. It gets its one-time password, then compresses the IP of the server, and the resulting string looks something like AzS9Sf4-purple-sausages.
There's still an infrastructure, yes, but it's peer-to-peer, it's self-configuring, it costs peanuts to run.
